packages/runtime/container-runtime/src/id-compressor/idCompressor.ts

```diff
@@ -545,1311 +239,424 @@ export class IdCompressor implements IIdCompressorCore, IIdCompressor {
+			bufferUint: new BigUint64Array(buffer),
+		};
+		const version = readNumber(index);
+		assert(version === currentWrittenVersion, "Unknown serialized version.");
+		const hasLocalState = readBoolean(index);
+		const clusterCapacity = readNumber(index);
+		const sessionCount = readNumber(index);
+		const clusterCount = readNumber(index);
+
+		// Sessions
+		let sessionOffset: number;
+		let sessions: [NumericUuid, Session][];
+		if (hasLocalState) {
+			sessionOffset = 0;
+			sessions = [];
 		} else {
-			localSessionId = newSessionIdMaybe;
+			// If !hasLocalState, there won't be a serialized local session ID so insert one at the beginning
+			assert(sessionId !== undefined, "Local session ID is undefined.");
+			const localSessionNumeric = numericUuidFromStableId(sessionId);
+			sessions = [[localSessionNumeric, new Session(localSessionNumeric)]];
+			sessionOffset = 1;
 		}
-
-		const compressor = new IdCompressor(localSessionId);
-		compressor.clusterCapacity = clusterCapacity;
-
-		const localOverridesInverse = new Map<string, LocalCompressedId>();
-		if (serializedLocalState !== undefined) {
-			// Do this part of local rehydration first since the cluster map population needs to query to local overrides
-			compressor.localIdCount = serializedLocalState.localIdCount;
-			compressor.lastTakenLocalId = serializedLocalState.lastTakenLocalId;
-			if (serializedLocalState.overrides !== undefined) {
-				for (const [localId, override] of serializedLocalState.overrides) {
-					compressor.localOverrides.append(localId, override);
-					localOverridesInverse.set(override, localId);
-					compressor.clustersAndOverridesInversion.set(
-						IdCompressor.createInversionKey(override),
-						localId as UnackedLocalId,
-					);
-				}
-			}
-		}
-
-		const sessionInfos: {
-			readonly session: Session;
-			readonly sessionId: SessionId;
-		}[] = [];
-		for (const serializedSession of serializedSessions) {
-			const [sessionId] = serializedSession;
-			if (sessionId === localSessionId) {
-				assert(hasOngoingSession(serialized), 0x499 /* Cannot resume existing session. */);
-				sessionInfos.push({ session: compressor.localSession, sessionId });
-			} else {
-				const session = compressor.createSession(sessionId);
-				sessionInfos.push({ session, sessionId });
-			}
+		for (let i = 0; i < sessionCount; i++) {
+			const numeric = readNumericUuid(index);
+			sessions.push([numeric, new Session(numeric)]);
 		}
 
-		for (const serializedCluster of serializedClusters) {
-			const { sessionIndex, capacity, count, overrides } =
-				deserializeCluster(serializedCluster);
-			const { session, sessionId } = sessionInfos[sessionIndex];
-			const { lastFinalizedLocalId, sessionUuid } = session;
-			const currentIdCount = lastFinalizedLocalId === undefined ? 0 : -lastFinalizedLocalId;
+		const compressor = new IdCompressor({
+			sessions: new Sessions(sessions),
+			localSessionId: stableIdFromNumericUuid(sessions[0][0]) as SessionId,
```

Comment:
You can avoid having the IdCompressor take this param because it already assumes that the first session in order is the local session